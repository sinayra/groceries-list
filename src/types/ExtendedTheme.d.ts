export interface ExtendedTheme extends Theme {
    dark: boolean,
    colors: {
      primary: string,
      background: string,
      card: string,
      text: string,
      border: string,
      notification: string,
      red: string,
      green: string,
      yellow: string,
      secondaryCard: string,
    }
  }