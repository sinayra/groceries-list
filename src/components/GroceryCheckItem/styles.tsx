import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
    container: {
        flexDirection: "row",
        justifyContent: "space-between",
        flex: 1,
    },
    checkbox: {
        flexDirection: "row",
        alignItems: "center",
        flex: 3,
    },
    title: {
        fontFamily: "Ubuntu_700Bold",
    },
    textItem: {
        flexDirection: "row",
    },
    quantity: {
        flexDirection: "row",
    },
    quantityText: {
        paddingLeft: 10,
        paddingRight: 10,
    },
    price: {
        paddingLeft: 20,
        flex: 1,
    }
});

export default styles;