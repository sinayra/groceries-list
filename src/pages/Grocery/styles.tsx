import { StyleSheet } from "react-native";
import Constants from "expo-constants";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    flexDirection: "column",
    paddingTop: 20 + Constants.statusBarHeight,
    paddingBottom: 20 + Constants.statusBarHeight,
    paddingLeft: 10,
    paddingRight: 10,
    justifyContent: "space-between",
    alignItems: "stretch",
    alignContent: "stretch",
  },
  content: {
    backgroundColor: "#FFA",
    flex: 1,
  },
  button: {
    backgroundColor: "#34CB79",
    height: 60,
    flexDirection: "row",
    borderRadius: 10,
    overflow: "hidden",
    alignItems: "center",
    marginTop: 8,
  },
  buttonText: {
    flex: 1,
    justifyContent: "center",
    textAlign: "center",
    color: "#FFFFFF",
    fontFamily: "Roboto_500Medium",
    fontSize: 16,
  },
  input: {
    height: 54,
    backgroundColor: "#FFF",
    justifyContent: "center",
    paddingHorizontal: 16,
    marginTop: 16,
    borderColor: '#FFF',
    borderBottomColor: '#CCC',
    borderWidth: 2,
  },
  label: {
    fontFamily: 'Roboto_400Regular',
    color: '#D4C2FF',
  },
  inputGroup: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  inputBlock: {
    width: '48%',
  },
  form: {
    flex: 1,
  },
  showNew: {
    flexDirection: 'row',
    alignItems: 'center'
  }
});

export default styles;
