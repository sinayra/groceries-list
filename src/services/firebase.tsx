import firebase from "firebase";
import {
  apiKey,
  authDomain,
  databaseURL,
  projectId,
  storageBucket,
  messagingSenderId,
  appId,
  measurementId,
} from "../config/keys.json";

const firebaseConfig = {
  apiKey,
  authDomain,
  databaseURL,
  projectId,
  storageBucket,
  messagingSenderId,
  appId,
  measurementId,
};

firebase.initializeApp(firebaseConfig);

export default firebase;
