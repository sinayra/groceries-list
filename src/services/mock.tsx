const mock = {
  "groceries" : {
    "0" : {
      "name" : "Coca 1L",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 11.18,
        "quantity" : 2
      }, {
        "date" : 1599955200000,
        "price" : 10.38,
        "quantity" : 2
      }, {
        "date" : 1600041600000,
        "price" : 10.38,
        "quantity" : 2
      }, {
        "date" : 1600646400000,
        "price" : 11.18,
        "quantity" : 2
      } ]
    },
    "1" : {
      "name" : "Coca Zero 1.5L",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 7.78,
        "quantity" : 2
      }, {
        "date" : 1600646400000,
        "price" : 3.89,
        "quantity" : 1
      } ]
    },
    "2" : {
      "name" : "Chá",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 8.99,
        "quantity" : 1
      }, {
        "date" : 1600646400000,
        "price" : 17.98,
        "quantity" : 2
      } ]
    },
    "3" : {
      "name" : "Sabonete",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 1.34,
        "quantity" : 1
      } ]
    },
    "4" : {
      "name" : "Batata",
      "purchases" : [ {
        "date" : 1594252800000,
        "price" : 2.77,
        "quantity" : 0.5551102204408818
      }, {
        "date" : 1595635200000,
        "price" : 1.4,
        "quantity" : 0.280561122244489
      }, {
        "date" : 1600041600000,
        "price" : 0.87,
        "quantity" : 0.2909698996655518
      }, {
        "date" : 1601078400000,
        "price" : 1.32,
        "quantity" : 0.44
      } ]
    },
    "5" : {
      "name" : "Cenoura",
      "purchases" : [ {
        "date" : 1595635200000,
        "price" : 1.0,
        "quantity" : 0.20040080160320642
      }, {
        "date" : 1595894400000,
        "price" : 0.81,
        "quantity" : 0.2903225806451613
      }, {
        "date" : 1600041600000,
        "price" : 0.74,
        "quantity" : 0.1544885177453027
      } ]
    },
    "6" : {
      "name" : "Pão de forma",
      "purchases" : [ {
        "date" : 1599264000000,
        "price" : 3.99,
        "quantity" : 1
      }, {
        "date" : 1600646400000,
        "price" : 3.19,
        "quantity" : 1
      }, {
        "date" : 1598313600000,
        "price" : 6.07,
        "quantity" : 1
      }, {
        "date" : 1593734400000,
        "price" : 4.39,
        "quantity" : 1
      }, {
        "date" : 1595116800000,
        "price" : 3.4,
        "quantity" : 1
      }, {
        "date" : 1591142400000,
        "price" : 4.19,
        "quantity" : 1
      }, {
        "date" : 1591401600000,
        "price" : 4.29,
        "quantity" : 1
      }, {
        "date" : 1592092800000,
        "price" : 4.79,
        "quantity" : 1
      }, {
        "date" : 1593043200000,
        "price" : 7.19,
        "quantity" : 1
      } ]
    },
    "-MJ7uyt7ZxyrE_XkrdQo" : {
      "name" : "mais um produto",
      "purchases" : {
        "-MJ7uyu6vHZ-hgRS8Lna" : {
          "date" : 1602172468142,
          "price" : 3,
          "quantity" : 3
        },
        "-MJ7v2o4EbJaTO90QG-T" : {
          "date" : 1602172468142,
          "price" : 2,
          "quantity" : 2
        }
      }
    }
  }
}

export default mock;
