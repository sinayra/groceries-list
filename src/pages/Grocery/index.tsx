import React from "react";
import {
  View,
  Text,
  TouchableOpacity,
  SafeAreaView,
  Linking,
} from "react-native";
import { RectButton } from "react-native-gesture-handler";
import styles from "./styles";

export default function Home() {
  function handleSavePurchase() {
    console.info("not implemented");
  }

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.content}>
        <View style={styles.form}>
          <Text>Grocery page</Text>
        </View>
        <RectButton style={styles.button} onPress={handleSavePurchase}>
          <Text style={styles.buttonText}>Salvar</Text>
        </RectButton>
      </View>
    </SafeAreaView>
  );
}
