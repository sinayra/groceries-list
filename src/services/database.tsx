import firebase from "./firebase";
import { Grocery, Purchase } from "../types/Grocery";

const database = firebase.database();

export async function getGroceries(): Promise<Grocery[]> {
  const promise: Promise<Grocery[]> = new Promise((resolve, reject) => {
    database.ref("/groceries").orderByChild("name").on("value", (snapshot) => {
      let groceries: Grocery[] = [];

      snapshot.forEach((elemSnapshot) => {
        let purchases: Purchase[] = [];
        const id = elemSnapshot.key ? elemSnapshot.key : undefined;
        const name = elemSnapshot.child("name").val();
        const obj = elemSnapshot.child("purchases");

        obj.forEach((purchasesSnapshot) => {
          const id = purchasesSnapshot.key ? purchasesSnapshot.key : undefined;
          const date = purchasesSnapshot.child("date").val();
          const price = purchasesSnapshot.child("price").val();
          const quantity = purchasesSnapshot.child("quantity").val();

          purchases.push({ id, date, price, quantity });
        });

        groceries.push({ id, name, purchases });
      });

      resolve(groceries);
    });
  });

  return promise;
}

export async function getPurchases(id: string): Promise<Purchase[]> {
  const promise: Promise<Purchase[]> = new Promise((resolve, reject) => {
    database.ref(`/groceries/${id}/purchases/`).on("value", (snapshot) => {
      let purchases: Purchase[] = [];

      snapshot.forEach((elemSnapshot) => {
        const id = elemSnapshot.key ? elemSnapshot.key : undefined;
        const date = elemSnapshot.child("date").val();
        const price = elemSnapshot.child("price").val();
        const quantity = elemSnapshot.child("quantity").val();

        purchases.push({ id, date, price, quantity });
      });

      resolve(purchases);
    });
  });

  return promise;
}

export async function setGroceries(
  id: string | undefined,
  name: string,
  date: number,
  price: number,
  quantity: number
) {
  let grocery: string | null | undefined = id;

  if (grocery === undefined) {
    grocery = (await database.ref("/groceries").push()).key;

    if (grocery !== null) {
      database.ref(`groceries/${grocery}`).set({
        name,
        purchases: {},
      });
    }
  }

  if (grocery !== null) {
    const newPurchase = (await database.ref(`groceries/${grocery}`).push()).key;

    database.ref(`groceries/${grocery}/purchases/${newPurchase}`).set({
      date,
      price,
      quantity,
    });

    return 200;
  } else {
    return 500;
  }
}

export async function deleteGroceries(
  idGrocery: string,
  idPurchase: string | null = null
) {
  if (idPurchase === null) {
    await database.ref(`/groceries/${idGrocery}`).remove();
  } else {
    await database
      .ref(`/groceries/${idGrocery}/purchases/${idPurchase}`)
      .remove();
  }

  return 200;
}

export async function getPurchaseList(): Promise<Grocery[]> {
  const promise: Promise<Grocery[]> = new Promise((resolve, reject) => {
    database.ref("/list").on("value", (snapshot) => {
      let groceries: Grocery[] = [];
      console.log(snapshot);

      snapshot.forEach((listId) => {

        const id = listId.key ? listId.key : undefined;
        const groceryId = listId.child("id").val();
        console.log("ID", id)

        database.ref(`/groceries/${groceryId}`).on("value", (grocery) => {
          console.log("GROCERY", grocery)
          let purchases: Purchase[] = [];
          const name = grocery.child("name").val();
          const obj = grocery.child("purchases");

          obj.forEach((purchasesSnapshot) => {
            const purchaseId = purchasesSnapshot.key ? purchasesSnapshot.key : undefined;
            const date = purchasesSnapshot.child("date").val();
            const price = purchasesSnapshot.child("price").val();
            const quantity = purchasesSnapshot.child("quantity").val();

            purchases.push({ id: purchaseId, date, price, quantity });
          });

          groceries.push({ listId: id, id: groceryId, name, purchases });
        });


      });

      resolve(groceries);
    });

  });

  return promise;
}

export async function addToPurchaseList(
  id: string | undefined,
) {
  if (id) {
    const key = (await database.ref(`/list/${id}`).push()).key;
    database.ref(`list/${key}`).set({
      id
    });
    return 200;
  } else {
    return 500;
  }
}

export async function removeFromPurchaseList(
  id: string | undefined,
) {
  if (id) {
    await database.ref(`/list/${id}`).remove();

    return 200;
  } else {
    return 500;
  }
}
