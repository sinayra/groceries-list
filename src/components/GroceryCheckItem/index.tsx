import React, { useEffect, useState } from "react";
import { View, Text } from "react-native";
import CheckBox from '@react-native-community/checkbox';
import { FontAwesome } from '@expo/vector-icons';
import { TouchableOpacity } from "react-native-gesture-handler";

import { addToPurchaseList, removeFromPurchaseList, setQuantity } from "../../services/database";
import { createPaidPriceArray, calculateThresholdMode } from "../../utils/purchaseMath";

import styles from "./styles";

import { Variables } from "../../styles/variables";
import { Grocery } from "../../types/Grocery";

interface GroceryCheckItemProps {
    item: Grocery;
    canBeAddedToList: boolean;
    reload: () => void;
}

const GroceryCheckItem: React.FC<GroceryCheckItemProps> = ({ item, canBeAddedToList, reload }) => {
    const variables = Variables();
    const [check] = useState(item.listId ? true : false);
    const [price, setPrice] = useState(0);

    const { name } = item;

    function calculatePrice(quantity?: number) {
        const purchaseHistory = createPaidPriceArray(item.purchases);
        const mode = calculateThresholdMode(purchaseHistory, 0.1);

        if (mode) {
            if (quantity && quantity > 0) {
                setPrice(mode * quantity);
            }
            else {
                setPrice(mode);
            }
        }
    }

    useEffect(() => {
        calculatePrice(item.listQuantity);
    }, []);


    async function handleAddToList(value: boolean) {
        if (value === true) {
            await addToPurchaseList(item.id);
        }
        else {
            await removeFromPurchaseList(item.listId);
        }

        reload();
    }

    function handleQuantity(increment: boolean) {
        if (item.listQuantity) {
            if (increment) {
                setQuantity(item.listId, item.listQuantity + 1);
                calculatePrice(item.listQuantity + 1)
            }
            else {
                if (item.listQuantity > 1) {
                    setQuantity(item.listId, item.listQuantity - 1);
                    calculatePrice(item.listQuantity - 1)
                }
                else {
                    removeFromPurchaseList(item.listId);
                }
            }
            reload();
        }
    }

    return (
        <View style={{ ...styles.container, opacity: canBeAddedToList ? 0.5 : 1 }}>
            <View style={styles.checkbox}>
                <CheckBox
                    disabled={false}
                    value={check}
                    onValueChange={(value) => handleAddToList(value)}
                />
                <View style={styles.textItem}>
                    <View>
                        <Text
                            style={{
                                ...styles.title,
                                color: variables.TEXT_COLOR,
                                fontSize: variables.FONT_SIZE_LARGE,
                            }}
                        >
                            {name}

                        </Text>
                    </View>
                </View>
            </View>

            {canBeAddedToList === false &&
                <View style={styles.quantity} testID="cannotBeAddToList">
                    <TouchableOpacity onPress={() => handleQuantity(true)} testID="add">
                        <FontAwesome name="plus-circle" size={variables.FONT_SIZE_LARGE + 10} color={variables.TEXT_COLOR} />
                    </TouchableOpacity>

                    <Text style={styles.quantityText}>x{item.listQuantity}</Text>

                    <TouchableOpacity onPress={() => handleQuantity(false)} testID="sub">
                        <FontAwesome name="minus-circle" size={variables.FONT_SIZE_LARGE + 10} color={variables.TEXT_COLOR} />
                    </TouchableOpacity>
                </View>
            }

            <View style={styles.price} testID="price">
                <Text style={{ fontSize: variables.FONT_SIZE_LARGE, color: variables.TEXT_COLOR, opacity: 0.5 }}>+ â‚¬{price.toFixed(2)}</Text>
            </View>

        </View>
    );
};

export default GroceryCheckItem;
