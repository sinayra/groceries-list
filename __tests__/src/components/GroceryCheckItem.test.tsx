import React from 'react';
import { shallow } from 'enzyme';
import { Text } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import CheckBox from '@react-native-community/checkbox';

import GroceryCheckItem from "../../../src/components/GroceryCheckItem";
import { Grocery } from "../../../src/types/Grocery";
import * as database from "../../../src/services/database";
import { TouchableOpacity } from 'react-native-gesture-handler';

describe('<Grocery Check Item />', () => {
  let mockAdd: jest.SpyInstance<Promise<200 | 500>, [id: string | undefined]>;
  let mockRemove: jest.SpyInstance<Promise<200 | 500>, [idList: string | undefined]>;
  let mockQuantity: jest.SpyInstance<Promise<200 | 500>, [idList: string | undefined, quantity: number]>;

  beforeEach(() => {
    mockAdd = jest.spyOn(database, 'addToPurchaseList').mockImplementation(jest.fn());
    mockRemove = jest.spyOn(database, 'removeFromPurchaseList').mockImplementation(jest.fn());
    mockQuantity = jest.spyOn(database, 'setQuantity').mockImplementation(jest.fn());
  });

  afterEach(() => {
    mockRemove.mockRestore();
    mockAdd.mockRestore();
    mockQuantity.mockRestore();
  });

  describe('Item in the purchase list', () => {
    let reload: () => void;
    let addToList: boolean;

    beforeAll(() => {
      reload = jest.fn();
      addToList = false;
    });

    it('Load text', async () => {
      const item: Grocery = {
        name: "Sabonete",
        listId: "0",
        listQuantity: 1,
        purchases: [{
          "date": 1599264000000,
          "price": 1.34,
          "quantity": 1
        }]
      }

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);

      const text = component.find(Text).getElements();

      expect(text).toHaveLength(3);

      const title = text[0].props.children;
      const quantity = text[1].props.children;
      const price = text[2].props.children;

      expect(title).toContain(item.name);
      expect(quantity).toContain(item.purchases[0].quantity);
      expect(price).toContain(item.purchases[0].price.toFixed(2));
    });

    it('Calculate new price based on list quantity', async () => {
      const item: Grocery = {
        name: "Sabonete",
        listId: "0",
        listQuantity: 2,
        purchases: [{
          "date": 1599264000000,
          "price": 1.34,
          "quantity": 1
        }]
      }

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);

      const text = component.find(Text).getElements();

      expect(text).toHaveLength(3);

      const title = text[0].props.children;
      const quantity = text[1].props.children;
      const price = text[2].props.children;

      expect(title).toContain(item.name);
      expect(quantity).toContain(item.listQuantity);
      expect(price).toContain("2.68");
    });

    it('Empty purchases', async () => {
      const item: Grocery = {
        name: "Sabonete",
        listId: "0",
        listQuantity: 1,
        purchases: []
      }

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);

      const text = component.find(Text).getElements();

      expect(text).toHaveLength(3);

      const title = text[0].props.children;
      const quantity = text[1].props.children;
      const price = text[2].props.children;

      expect(title).toContain(item.name);
      expect(quantity).toContain(item.listQuantity);
      expect(price).toContain("0.00");
    });

    it('Handle Add to List', async () => {
      const item: Grocery = {
        name: "Sabonete",
        listId: "0",
        listQuantity: 1,
        purchases: [{
          "date": 1599264000000,
          "price": 1.34,
          "quantity": 1
        }]
      }

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);
      const checkbox = component.find(CheckBox);

      const handler = checkbox.prop("onValueChange");
      const value = checkbox.prop("value");

      if (handler && value !== undefined) {
        handler(!value);
      }

      expect(mockRemove).toBeCalled();
      expect(mockAdd).not.toBeCalled();

    });

    describe('Handle quantity', () => {
      it('loads quantity icons', async () => {
        const item: Grocery = {
          name: "Sabonete",
          listId: "0",
          listQuantity: 1,
          purchases: [{
            "date": 1599264000000,
            "price": 1.34,
            "quantity": 1
          }]
        };

        const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />).find({ testID: "cannotBeAddToList" });
        const icons = component.find(FontAwesome).getElements();
        const quantity = component.find(Text).getElement().props.children;

        expect(icons).toHaveLength(2);

        const add = icons[0].props.name;
        const sub = icons[1].props.name;

        expect(add).toEqual("plus-circle");
        expect(sub).toEqual("minus-circle");
        expect(quantity).toContain(1);
      });

      it('add quantity', async () => {
        const item: Grocery = {
          name: "Sabonete",
          listId: "0",
          listQuantity: 1,
          purchases: [{
            "date": 1599264000000,
            "price": 1.34,
            "quantity": 1
          }]
        };

        const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);
        const icons = component.find({ testID: "cannotBeAddToList" }).find(TouchableOpacity);
        const add = icons.find({ testID: "add" });

        const handler = add.prop("onPress");

        if (handler !== undefined) {
          handler(true);
        }

        const price = component.find({ testID: "price" }).find(Text).getElement().props.children;

        expect(mockQuantity).toBeCalled();
        expect(price).toContain((2.68).toFixed(2));
      });

      describe('subtract quantity', () => {
        it('multiple elements', async () => {
          const item: Grocery = {
            name: "Sabonete",
            listId: "0",
            listQuantity: 5,
            purchases: [{
              "date": 1599264000000,
              "price": 1.34,
              "quantity": 1
            }]
          };

          const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);
          const icons = component.find({ testID: "cannotBeAddToList" }).find(TouchableOpacity);
          const sub = icons.find({ testID: "sub" });

          const handler = sub.prop("onPress");

          if (handler !== undefined) {
            handler(false);
          }

          const price = component.find({ testID: "price" }).find(Text).getElement().props.children;

          expect(mockQuantity).toBeCalled();
          expect(price).toContain((5.36).toFixed(2));
        });

        it('one element', async () => {
          const item: Grocery = {
            name: "Sabonete",
            listId: "0",
            listQuantity: 1,
            purchases: [{
              "date": 1599264000000,
              "price": 1.34,
              "quantity": 1
            }]
          };

          const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);
          const icons = component.find({ testID: "cannotBeAddToList" }).find(TouchableOpacity);
          const sub = icons.find({ testID: "sub" });

          const handler = sub.prop("onPress");

          if (handler !== undefined) {
            handler(false);
          }

          expect(mockRemove).toBeCalled();
        });
      });
    });

  });

  describe('Item not in the purchase list', () => {
    let reload: () => void;
    let addToList: boolean;

    beforeAll(() => {
      reload = jest.fn();
      addToList = true;
    });

    it('Load text', async () => {
      const item: Grocery = {
        name: "Sabonete",
        purchases: [{
          "date": 1599264000000,
          "price": 1.34,
          "quantity": 1
        }]
      };

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);

      const text = component.find(Text).getElements();

      expect(text).toHaveLength(2);

      const title = text[0].props.children;
      const price = text[1].props.children;

      expect(title).toContain(item.name);
      expect(price).toContain(item.purchases[0].price.toFixed(2));
    });

    it('Empty purchases', async () => {
      const item: Grocery = {
        name: "Sabonete",
        purchases: []
      };

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);
      const text = component.find(Text).getElements();

      expect(text).toHaveLength(2);

      const title = text[0].props.children;
      const price = text[1].props.children;

      expect(title).toContain(item.name);
      expect(price).toContain("0.00");
    });

    it('Handle Add to List', async () => {
      const item: Grocery = {
        name: "Sabonete",
        purchases: [{
          "date": 1599264000000,
          "price": 1.34,
          "quantity": 1
        }]
      }

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />);
      const checkbox = component.find(CheckBox);

      const handler = checkbox.prop("onValueChange");
      const value = checkbox.prop("value");

      if (handler && value !== undefined) {
        handler(!value);
      }

      expect(mockAdd).toBeCalled();
      expect(mockRemove).not.toBeCalled();

    });

    it('Does not loads quantity icons', async () => {
      const item: Grocery = {
        name: "Sabonete",
        purchases: [{
          "date": 1599264000000,
          "price": 1.34,
          "quantity": 1
        }]
      };

      const component = shallow(<GroceryCheckItem item={item} canBeAddedToList={addToList} reload={reload} />).find({ testID: "cannotBeAddToList" });
      const icons = component.find(FontAwesome).getElements();

      expect(icons).toHaveLength(0);
    });

  });

});
