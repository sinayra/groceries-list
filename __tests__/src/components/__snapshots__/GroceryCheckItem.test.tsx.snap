// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grocery Check Item /> Item in the purchase list Snapshot 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 1,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 3,
        "flexDirection": "row",
      }
    }
  >
    <ForwardRef(CheckBoxWithRef)
      disabled={false}
      onValueChange={[Function]}
      value={true}
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
      testID="item-name"
    >
      <Component>
        <Component
          style={
            Object {
              "color": "rgb(28, 28, 30)",
              "fontFamily": "Ubuntu_700Bold",
              "fontSize": 16,
            }
          }
        >
          Sabonete
        </Component>
      </Component>
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flexDirection": "row",
      }
    }
    testID="cannotBeAddToList"
  >
    <TouchableOpacity
      activeOpacity={0.2}
      delayLongPress={600}
      extraButtonProps={
        Object {
          "rippleColor": "transparent",
        }
      }
      onPress={[Function]}
      testID="add"
    >
      <Icon
        allowFontScaling={false}
        color="rgb(28, 28, 30)"
        name="plus-circle"
        size={26}
      />
    </TouchableOpacity>
    <Component
      style={
        Object {
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      x
      1
    </Component>
    <TouchableOpacity
      activeOpacity={0.2}
      delayLongPress={600}
      extraButtonProps={
        Object {
          "rippleColor": "transparent",
        }
      }
      onPress={[Function]}
      testID="sub"
    >
      <Icon
        allowFontScaling={false}
        color="rgb(28, 28, 30)"
        name="minus-circle"
        size={26}
      />
    </TouchableOpacity>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
        "paddingLeft": 20,
      }
    }
    testID="price"
  >
    <Component
      style={
        Object {
          "color": "rgb(28, 28, 30)",
          "fontSize": 16,
          "opacity": 0.5,
        }
      }
    >
      + €
      1.34
    </Component>
  </Component>
</Component>
`;

exports[`<Grocery Check Item /> Item not in the purchase list Snapshot 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "opacity": 0.5,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 3,
        "flexDirection": "row",
      }
    }
  >
    <ForwardRef(CheckBoxWithRef)
      disabled={false}
      onValueChange={[Function]}
      value={false}
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
        }
      }
      testID="item-name"
    >
      <Component>
        <Component
          style={
            Object {
              "color": "rgb(28, 28, 30)",
              "fontFamily": "Ubuntu_700Bold",
              "fontSize": 16,
            }
          }
        >
          Sabonete
        </Component>
      </Component>
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
        "paddingLeft": 20,
      }
    }
    testID="price"
  >
    <Component
      style={
        Object {
          "color": "rgb(28, 28, 30)",
          "fontSize": 16,
          "opacity": 0.5,
        }
      }
    >
      + €
      1.34
    </Component>
  </Component>
</Component>
`;
